cpp/src/parquet/column_reader_test.cc

```diff
@@ -572,5 +576,615 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+    std::vector<int16_t> read_defs(
+        record_reader->def_levels(),
+        record_reader->def_levels() + record_reader->levels_position());
+    std::vector<int16_t> read_reps(
+        record_reader->rep_levels(),
+        record_reader->rep_levels() + record_reader->levels_position());
+
+    ASSERT_THAT(read_vals, ElementsAre(20, 20, 20, 40));
+    ASSERT_THAT(read_defs, ElementsAre(1, 1, 1, 1));
+    ASSERT_THAT(read_reps, ElementsAre(0, 1, 1, 0));
+  }
+}
+
+// Test reading when one record spans multiple pages for a repeated field.
+TEST(RecordReaderTest, ReadPartialRecord) {
+  internal::LevelInfo level_info;
+  level_info.def_level = 1;
+  level_info.rep_level = 1;
+
+  NodePtr type = schema::Int32("b", Repetition::REPEATED);
+  const ColumnDescriptor descr(type, level_info.def_level, level_info.rep_level);
+
+  std::vector<std::shared_ptr<Page>> pages;
+  std::unique_ptr<PageReader> pager;
+
+  // Page 1: {[10], [20, 20, 20 ... } continues to next page.
+  {
+    std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+        &descr, /*values=*/{10, 20, 20, 20}, /*num_values=*/4, Encoding::PLAIN,
+        /*indices=*/{},
+        /*indices_size=*/0, /*def_levels=*/{1, 1, 1, 1}, level_info.def_level,
+        /*rep_levels=*/{0, 0, 1, 1}, level_info.rep_level);
+    pages.push_back(std::move(page));
+  }
+
+  // Page 2: {... 20, 20, ...} continues from previous page and to next page.
+  {
+    std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+        &descr, /*values=*/{20, 20}, /*num_values=*/2, Encoding::PLAIN,
+        /*indices=*/{},
+        /*indices_size=*/0, /*def_levels=*/{1, 1}, level_info.def_level,
+        /*rep_levels=*/{1, 1}, level_info.rep_level);
+    pages.push_back(std::move(page));
+  }
+
+  // Page 3: { ... 20, [30]} continues from previous page.
+  {
+    std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+        &descr, /*values=*/{20, 30}, /*num_values=*/2, Encoding::PLAIN,
+        /*indices=*/{},
+        /*indices_size=*/0, /*def_levels=*/{1, 1}, level_info.def_level,
+        /*rep_levels=*/{1, 0}, level_info.rep_level);
+    pages.push_back(std::move(page));
+  }
+
+  pager.reset(new test::MockPageReader(pages));
+
+  std::shared_ptr<internal::RecordReader> record_reader =
+      internal::RecordReader::Make(&descr, level_info);
+  record_reader->SetPageReader(std::move(pager));
+
+  {
+    // Read [10]
+    int64_t records_read = record_reader->ReadRecords(/*num_records=*/1);
+
+    ASSERT_EQ(records_read, 1);
+
+    const auto read_values = reinterpret_cast<const int32_t*>(record_reader->values());
+    std::vector<int32_t> read_vals(read_values,
+                                   read_values + record_reader->values_written());
+
```

Comment:
Why aren't we also testing `def_levels()` and `rep_levels()` here?

Reply:
Also, same question about `null_count()`, `levels_written()` etc.

Reply:
Done.